### ISSUE TITLE
Question about a method in the spatial pooler and temporal memory in algorithm 

### ISSUE THREAD

**Issue by zengwufu** [htm-community/htm.core#929]
from htm.bindings.algorithms import SpatialPooler
from htm.bindings.algorithms import TemporalMemory

these two class both have  a similar method named loadFromFile(...) and  saveToFile(...) , and these two methods has a parameter named args0: str, what is the meaing of it?   is it file path string or something else ? 

i try to use it as file path string, but it does'n work. i run the anomaly detect program using the sp and tm, and i save the tm use the method--saveToFile(file_path_string). then i write a test methods to load the tm from the file and print the tm's getActiveCells(), there is nothing in the stdout. 

here is my code:
    anomaly detecting code:
      for count, record in enumerate(records[10000:]):
          if count >= numRecords: break
          
          dateString = datetime.datetime.strptime(record[0], "%Y-%m-%d %H:%M:%S")
          value = float(record[1])
          value2 = float(record[2])
          values.append(value)
          value2s.append(value2)
  
          dateBits = dateEncoder.encode(dateString)
          valueBits = valueEncoder.encode(value)
          vlaue2Bits = value2Encoder.encode(value2)
  
          activeColumnsDate = SDR(spDate.getColumnDimensions())
          activeColumnsValue = SDR(spValue.getColumnDimensions())
          activeColumnsValue2 = SDR(spValue2.getColumnDimensions())
  
          spDate.compute(dateBits, True, activeColumnsDate)
          spValue.compute(valueBits, True, activeColumnsValue)
          spValue2.compute(vlaue2Bits, True, activeColumnsValue2)
  
          activeColumns = SDR([
                  spDate.getColumnDimensions()[0] 
                  + spValue.getColumnDimensions()[0] 
                  + spValue2.getColumnDimensions()[0]
                  ]).concatenate([activeColumnsValue, activeColumnsValue2, activeColumnsDate])
  
          tm.compute(activeColumns, learn=True)
  
          anomalyScores.append(tm.anomaly)
  
          anomalyLikelihood = anomaly_history.anomalyProbability( value+value2, tm.anomaly )
          anomalyProb.append(anomalyLikelihood)
  
          if tm.anomaly > 0.5 and anomalyLikelihood > 0.7:
              print(dateString, value, value2, int(value2 + value), tm.anomaly, anomalyLikelihood)
  
      # save the spatial_pollers to file
      spDate.saveToFile(os.path.join(_SP_SAVED_FILE_DIR, 'spDate'))
      spValue.saveToFile(os.path.join(_SP_SAVED_FILE_DIR, 'spValue'))
      spValue2.saveToFile(os.path.join(_SP_SAVED_FILE_DIR, 'spValue2'))
  
      # save the temporal_memory to file
      tm.saveToFile(os.path.join(_TM_SAVED_FILE_DIR, 'tm'))

  the test method:
    def test_the_saved_tm():
      print("hahaha")
  
      tm = TemporalMemory().loadFromFile(os.path.join(_TM_SAVED_FILE_DIR, 'tm'))
      print(tm.getActiveCells())

and in the test method, if i put the first line print statement to the last line,  there is even no "hahaha" in the stdout.

**Comment by dkeeney** [comment]
The saveToFile( ), save( ), loadFromFile( ), and load( ) are part of the serialization facility that allows the state of a currently running project to be saved to a file or stream and then later reloaded and continue execution where it left off.  You will find these functions defined in `src/htm/types/Serializable.h`.

It is intended that an application will call either SaveToFile( ) or save( ) and LoadFromFile( ) or load( ) only on the Network object.  This will then recursively call into all other modules in order to serialize or deserialize the current state.
```
 /**
   *    saveToFile(path [,fmt])          Open a file and stream to it.
   *    save(ostream f [, fmt])   Stream to your stream.  
   *    f << net;                 Output human readable text. 
   *
   *    loadFromFile(path [, fmt])
   *    load(istream f [, fmt])
   *
   * @path The filename into which to save/load the streamed serialization.
   * @f    The stream with which to save/load the serialization.
   * @fmt  Format: One of following from enum SerializableFormat
   *   BINARY   - A binary format which is the fastest but not portable between platforms (default).
   *   PORTABLE - Another Binary format, not quite as fast but is portable between platforms.
   *   JSON     - Human readable JSON text format. Slow.
   *   XML      - Human readable XML text format. Even slower.
   *
   * See Serializable base class for more details.
   */
```

Under the covers, this calls the C++ `Cereal` serialization package which does all of the hard work.

An application can call these functions without the `fmt` argument which defaults to BINARY.
If you are calling from C++ then the `fmt` argument is an enum.
If you are calling from Python, the `fmt` argument is a string containing "BINARY", "PORTABLE", "JSON", or "XML".
  (See bindings/py/pp_src/bindings/engine/py_Engine.cpp about line 385)

**Comment by Finnfj** [comment]
I have a similar problem with the storing/loading of the TM in python doesn't seem to work properly (or I'm doing sth very wrong).
The topic is here: https://discourse.numenta.org/t/question-about-serialization-behaviour/8503

> If you are calling from Python, the fmt argument is a string containing "BINARY", "PORTABLE", "JSON", or "XML".
In the python bindings it looks like it'll default to binary though and have no other options (htm.core/bindings/py/cpp_src/bindings/algorithms/py_TemporalMemory.cpp):
`py_HTM.def("saveToFile",
				    [](TemporalMemory &self, const std::string& filename) {self.saveToFile(filename,SerializableFormat::BINARY); });

        py_HTM.def("loadFromFile",
				    [](TemporalMemory &self, const std::string& filename) { return self.loadFromFile(filename,SerializableFormat::BINARY); });`

**Comment by dkeeney** [comment]
Oh, now I see what you are talking about.

If you were using the NetworkAPI then you would call saveToFile( ) and loadFromFile( ) on the Network object and it serializes/deserializes everything for you.  But if you are not using NetworkAPI, then you do need to serialize each component yourself.  And, in this case you are calling these functions on the TM object.  ....  and the interface for these functions do not include the second argument.   Good catch.   This problem may also be on some of the other components as well.

So, that is a design error and should be corrected.  I will work on adding the second argument to the bindings.  I will try to do this within the next few days.

**Comment by ctrl-z-9000-times** [comment]
Hi David,

Another design issue is that in python the `Classifier` and `Predictor` classes do not support serialization despite the fact that they learn and are stateful. Serialization is implemented in C++  but the Python bindings are missing (I never wrote them?)

**Comment by dkeeney** [comment]
Hi David,

As long as I am messing with serialization I guess I should take care of those two classes as well.

**Comment by ctrl-z-9000-times** [comment]
lol, all of these classes implement `pickle` ?!

I just checked and pickle appears to work fine for both the `Classifier` and `Predictor` classes, so we don't need to change any of the code.

**Comment by ctrl-z-9000-times** [comment]
The real problem here is that no one knew that these classes implemented the `pickle` protocol.
We should be advertising this feature!

Some background: `pickle` is a best-in-class solution for serialization in python. It is in the standard library. It does a really good job and about 99% of the time it just works. It interoperates across almost the entire python ecosystem.

**Comment by ctrl-z-9000-times** [comment]
I'm thinking of adding a note to the documentation of every python method which starts with: read/save/write/load.

The note would just say: "This class also works with the pickle module."

**Comment by zengwufu** [comment]
> Oh, now I see what you are talking about.
> 
> If you were using the NetworkAPI then you would call saveToFile( ) and loadFromFile( ) on the Network object and it serializes/deserializes everything for you. But if you are not using NetworkAPI, then you do need to serialize each component yourself. And, in this case you are calling these functions on the TM object. .... and the interface for these functions do not include the second argument. Good catch. This problem may also be on some of the other components as well.
> 
> So, that is a design error and should be corrected. I will work on adding the second argument to the bindings. I will try to do this within the next few days.

Thank you for your reply !!!
I will try to use the serializes/deserializes in NetworkAPI.

**Comment by dkeeney** [comment]
>> The note would just say: "This class also works with the pickle module."

So, you don't need to use NetworkAPI to just call pickle on your TM module from within Python.
What it does is use Cereal within the C++ and the binary data that results is passed to pickle when being serialized by pickle from python and it does the reverse when deserializing.  But it does not use the JSON or XML formats.

To get the JSON and XML formats so that it is readable, you would need to call the saveToFile( ) functions from Python and that is what I need to fix.

**Comment by zengwufu** [comment]
> > > The note would just say: "This class also works with the pickle module."
> 
> So, you don't need to use NetworkAPI to just call pickle on your TM module from within Python.
> What it does is use Cereal within the C++ and the binary data that results is passed to pickle when being serialized by pickle from python and it does the reverse when deserializing. But it does not use the JSON or XML formats.
> 
> To get the JSON and XML formats so that it is readable, you would need to call the saveToFile( ) functions from Python and that is what I need to fix.

yes. i just try to use Network API. I  run one half of data on the network, and network.saveToFile(), then i use network.saveToFile() to load network state and run the the another one half of data , it work very well.

**Comment by dkeeney** [comment]
Thank you.  Many hours were spent in getting Network API to do serialization right.

**Comment by dkeeney** [comment]
I started looking at this in more detail and I found that some of the algorithms (most notably Classifier) did not have unit tests or the test were disabled.  What I found was that neither pickle nor saveToFile() worked in some cases.
So I am reopening this issue and will be generating a new Pull request.

**Comment by dkeeney** [comment]
reopening
